src/ir/ir_traversing_visitor.h

```diff
@@ -53,56 +53,60 @@ class IRTraversingVisitor : public IRVisitor<Derived, Result> {
+  virtual Result PostVisit(CopyConst<IsConst, Module>& module, Result in_order_result) {
     return in_order_result;
   }
-  // Invoked before all the children of `block` is visited.
-  virtual Result PreVisit(const Block& block) {
+  // Invoked before all the children of `block` are visited.
+  virtual Result PreVisit(CopyConst<IsConst, Block>& block) {
     return GetDefaultValue();
   }
-  // Invoked after all the children of `block` is visited.
-  virtual Result PostVisit(const Block& block, Result in_order_result) {
+  // Invoked after all the children of `block` are visited.
+  virtual Result PostVisit(CopyConst<IsConst, Block>& block, Result in_order_result) {
     return in_order_result;
   }
-  // Invoked before all the children of `operation` is visited.
-  virtual Result PreVisit(const Operation& operation) {
+  // Invoked before all the children of `operation` are visited.
+  virtual Result PreVisit(CopyConst<IsConst, Operation>& operation) {
     return GetDefaultValue();
   }
-  // Invoked after all the children of `operation` is visited.
-  virtual Result PostVisit(const Operation& operation, Result in_order_result) {
+  // Invoked after all the children of `operation` are visited.
+  virtual Result PostVisit(CopyConst<IsConst, Operation>& operation,
+                           Result in_order_result) {
     return in_order_result;
   }
 
-  Result Visit(const Module& module) final override {
+  Result Visit(CopyConst<IsConst, Module>& module) final override {
     Result pre_visit_result = PreVisit(module);
     Result fold_result = common::utils::fold(
         module.blocks(), std::move(pre_visit_result),
-        [this](Result acc, const std::unique_ptr<Block>& block) {
+        [this](Result acc, CopyConst<IsConst, std::unique_ptr<Block>>& block) {
           return FoldResult(acc, block->Accept(*this));
         });
     return PostVisit(module, fold_result);
   }
 
-  Result Visit(const Block& block) final override {
+  Result Visit(CopyConst<IsConst, Block>& block) final override {
     Result pre_visit_result = PreVisit(block);
     Result fold_result = common::utils::fold(
         block.operations(), std::move(pre_visit_result),
-        [this](Result acc, const std::unique_ptr<Operation>& operation) {
+        [this](
+            Result acc,
+            CopyConst<IsConst, std::unique_ptr<Operation>>& operation) {
           return FoldResult(acc, operation->Accept(*this));
         });
     return PostVisit(block, fold_result);
   }
 
-  Result Visit(const Operation& operation) final override {
+  Result Visit(CopyConst<IsConst, Operation>& operation) final override {
     Result result = PreVisit(operation);
-    const Module* module = operation.impl_module();
+    auto* module = operation.impl_module();
     if (module != nullptr) {
-      result = FoldResult(std::move(result), module->Accept<Derived, Result>(*this));
+      result = FoldResult(result, module->Accept(*this));
```

Comment:
Why is `std::move` removed here? I believe that using `std::move` is correct: you are erasing the only name (`result`) pointing to the `PreVisit` result so that `Result` may be moved into `FoldResult` (and the result of that operation put back into `result`). If you're worried about the fact that the parameter may be const, I wouldn't: `std::move` does not activate the move constructor when used on a `const` name.

Reply:
Thanks Mark! I thought I explicitly disabled the copy constructor for one of the test result values (and enabled the move constructor) but I hadn't.
I've now done that and gotten rid of any potential for accidental copying AND gotten rid of any unnecessary moves (the removing the ones there now cause build failure).

Thanks for the catch